depot/io/gcs.py

```diff
@@ -0,0 +1,136 @@
+from typing import List
+import uuid
+from google.cloud import storage
+from depot.io import utils
+from depot.io.interfaces import FileStorage, StoredFile
+from depot._compat import unicode_text, percent_encode, percent_decode
+from google.cloud.exceptions import NotFound
+import datetime
+import os
+
+from depot.utils import make_content_disposition
+CANNED_ACL_PUBLIC_READ = 'public-read'
+CANNED_ACL_PRIVATE = 'private'
+
+class GCSStoredFile(StoredFile):
+    def __init__(self, file_id, blob):
+        self.blob = blob
+        self._fileid = file_id
+        self._closed = False
+
+        metadata = blob.metadata or {}
+        filename = metadata.get('x-depot-filename')
+        content_type = metadata.get('x-depot-content-type')
+
+        last_modified = blob.updated
+        content_length = blob.size
+
+        super(GCSStoredFile, self).__init__(file_id, filename, content_type, last_modified, content_length)
+        self._pos = 0
+
+    def read(self, n=-1):
+        if self.closed:
+            raise ValueError("I/O operation on closed file")
+
+        data = self.blob.download_as_bytes(start=self._pos, end=self._pos + n - 1 if n != -1 else None)
+        self._pos += len(data)
+        return data
+
+    def close(self, *args, **kwargs):
+        self._closed = True
+
+    @property
+    def closed(self):
+        return self._closed
+
+    @property
+    def public_url(self):
+        return self.blob.public_url
+
+
+class GCSStorage(FileStorage):
+    def __init__(self, project_id, credentials=None, bucket=None, policy=None, storage_class=None):
+        if credentials:
+            os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = credentials
```

Comment:
Why we are not passing the credentials to `storage.Client` when creating it?
Setting environment variables, especially from multiprocessing use cases. Also the env variable could be mutated from some other place and lead to effects at a distance that are hard to debug.